src/fs/file.rs

```diff
@@ -338,50 +335,57 @@ impl File {
     /// }
     /// ```
     ///
-    /// [`ErrorKind::Interrupted`]: std::io::ErrorKind::Interrupted
     /// [`ErrorKind::UnexpectedEof`]: std::io::ErrorKind::UnexpectedEof
-    pub async fn read_exact_at<T: IoBufMut>(
+    pub async fn read_exact_at<T>(&self, buf: T, pos: u64) -> crate::BufResult<(), T>
+    where
+        T: BoundedBufMut,
+    {
+        let orig_bounds = buf.bounds();
+        let (res, buf) = self.read_exact_slice_at(buf.slice(..), pos).await;
+        (res, T::from_buf_bounds(buf, orig_bounds))
+    }
+
+    async fn read_exact_slice_at<T: IoBufMut>(
         &self,
-        mut buf: T,
-        pos: u64,
+        mut buf: Slice<T>,
+        mut pos: u64,
     ) -> crate::BufResult<(), T> {
-        let buf_len = buf.bytes_total();
-
-        if pos.checked_add(buf_len as u64).is_none() {
+        if pos.checked_add(buf.bytes_total() as u64).is_none() {
             return (
                 Err(io::Error::new(
                     io::ErrorKind::InvalidInput,
                     "buffer too large for file",
                 )),
-                buf,
+                buf.into_inner(),
```

Comment:
It would result in _cleaner_ code, but I was semi-conscious of the micro-optimization impact: the fat `Slice` struct would be returned by the internal function (or rather, the `Future` impl after the async transform), only to be stripped to the inner buffer by the caller. With unified monomorphization between `T: IoBuf` and `Slice<T>`, 
it is likely that the call is not inlined and the return is done on stack.